src/server/node_recov_db.c

```diff
@@ -88,152 +83,189 @@
-	else
-		pdbnd->nd_name[0]=0;
+	strcpy(pdbnd->nd_name, pnode->nd_name);
+	strcpy(pnode->nd_savetm, pdbnd->nd_savetm);
 
-	/* node_index is used to sort vnodes upon recovery.
-	 * For Cray multi-MoM'd vnodes, we ensure that natural vnodes come
-	 * before the vnodes that it manages by introducing offsetting all
-	 * non-natural vnodes indices to come after natural vnodes.
+	savetype |= OBJ_SAVE_QS;
+	/* nodes do not have a qs area, so we cannot check whether qs changed or not 
+	 * hence for now, we always write the qs area, for now!
 	 */
+		
+	/* node_index is used to sort vnodes upon recovery.
+	* For Cray multi-MoM'd vnodes, we ensure that natural vnodes come
+	* before the vnodes that it manages by introducing offsetting all
+	* non-natural vnodes indices to come after natural vnodes.
+	*/
 	pdbnd->nd_index = (pnode->nd_nummoms * svr_totnodes) + pnode->nd_index;
 
 	if (pnode->nd_hostname)
 		strcpy(pdbnd->nd_hostname, pnode->nd_hostname);
-	else
-		pdbnd->nd_hostname[0]=0;
-
+	
 	if (pnode->nd_moms && pnode->nd_moms[0])
 		pdbnd->mom_modtime = pnode->nd_moms[0]->mi_modtime;
-	else
-		pdbnd->mom_modtime = 0;
-
+	
 	pdbnd->nd_ntype = pnode->nd_ntype;
 	pdbnd->nd_state = pnode->nd_state;
 	if (pnode->nd_pque)
 		strcpy(pdbnd->nd_pque, pnode->nd_pque->qu_qs.qu_name);
 	else
-		pdbnd->nd_pque[0]=0;
+		pdbnd->nd_pque[0] = 0;
 
-	/*
-	 * node attributes are saved in a different way than attributes of other objects
-	 * for other objects we directly call save_attr_db, but for node attributes
-	 * we massage some of the attributes. The special ones are pcpus
-	 * and resv_enable and sharing.
-	 */
-	CLEAR_HEAD(wrtattr);
-
-	for (j = 0; j < ND_ATR_LAST; ++j) {
-		/* skip certain ones: no-save and default values */
-		if ((node_attr_def[j].at_flags & ATR_DFLAG_NOSAVM) ||
-				(pnode->nd_attr[j].at_flags & ATR_VFLAG_DEFLT))
-			continue;
-
-		(void) node_attr_def[j].at_encode(&pnode->nd_attr[j],
-		                                  &wrtattr,
-		                                  node_attr_def[j].at_name,
-		                                  NULL,
-		                                  ATR_ENCODE_SVR,
-		                                  NULL);
-
-		node_attr_def[j].at_flags &= ~ATR_VFLAG_MODIFY;
-	}
+	if ((encode_attr_db(node_attr_def, pnode->nd_attr, ND_ATR_LAST, &pdbnd->cache_attr_list, &pdbnd->db_attr_list, 0)) != 0)
+		return -1;
 
-	vnode_sharing = (((pnode->nd_attr[ND_ATR_Sharing].at_flags & (ATR_VFLAG_SET | ATR_VFLAG_DEFLT))
-				== (ATR_VFLAG_SET | ATR_VFLAG_DEFLT))
-				&& ((pnode->nd_attr[ND_ATR_Sharing].at_val.at_long != VNS_UNSET)
-						&& (pnode->nd_attr[ND_ATR_Sharing].at_val.at_long != VNS_DFLT_SHARED)));
-	numattr = vnode_sharing;
-	psvrl = (svrattrl *)GET_NEXT(wrtattr);
-	while (psvrl) {
-		if ((strcmp(psvrl->al_name, ATTR_rescavail) == 0) &&
-				(strcmp(psvrl->al_resc, "ncpus") == 0)) {
+	/* MSTODO: how can we optimize this loop - eliminate this? */
+	psvrl = (svrattrl *) GET_NEXT(pdbnd->db_attr_list.attrs);
+	while (psvrl != NULL) {
```

Comment:
How about we create a new flag and mark the attributes that are not supposed to be encoded to DB? Then encode_attr_db can ignore such attributes and this loop can be avoided